Attach arrows to cards

eyuelt opened the pull request:
Now you can drag arrows around and dragging an arrow onto a card will connect them.

There's still work to be done, like being able to drag the tail (if we want to be able to do that) and expanding the click target because the arrow itself is really thin, but this handles the main functionality.

Some of the stuff is kind of messy and will be cleaned up in an upcoming PR. I split them up to make reviewing easier.

girum11 commented on src/client/app/graph/arrow/arrow.component.ts:
```diff
@@ -26,6 +30,13 @@ export class ArrowComponent implements OnInit {
 
+  // Whether or not tail dragging is in progress.
+  private tailDragging: boolean = false;
+  // Whether or not tip dragging is in progress.
```
(nit)

Add vertical spacing between each field, for visual clarity?

eyuelt replied:
Really? Seems clear to me. Plus vertical space aint cheap.

girum11 commented on src/client/app/graph/arrow/arrow.component.ts:
```diff
@@ -78,4 +89,70 @@ export class ArrowComponent implements OnInit {
+      this.tipPosition.y += newDragPnt.y - this.lastDragPnt.y;
+      const newTipPosition =
+        this.canvasService.viewportCoordToCanvasCoord(this.tipPosition);
+      this.arrow.tipPosition = {x: newTipPosition.x, y: newTipPosition.y};
```
Style these inline objects to match what you have above? (for readability, [96], [110])

eyuelt replied:
Done

girum11 commented on src/client/app/graph/arrow/arrow.component.ts:
```diff
@@ -78,4 +89,70 @@ export class ArrowComponent implements OnInit {
+        // TODO(eyuelt): reposition the arrow
+      }
+      this.tipDragging = false;
+      this.lastDragPnt = null;
```
(nit) avoid acronyms like "pnt"?

eyuelt replied:
Done

girum11 commented on src/client/app/graph/canvas/canvas.service.ts:
```diff
@@ -123,6 +128,33 @@ export class CanvasService {
 
+  getCardById(id: string): Card {
+    // TODO(eyuelt): change CollaborativeList to CollaborativeMap
+    let cardsArray = this.cards.asArray();
```
Seems like you have "let" here and "const" elsewhere. As a team, we should probably converge on what we like best.

Google's ES6 style guide says to use "const" everywhere possible, but I don't like how it semantically reads (versus, say, "let" and "var" from Swift lang). 

Which do you prefer? "let" or "const" by default?

eyuelt replied:
Good point. I prefer using "const" wherever possible, and using "let" whenever the variable isn't const. Same as the Google C++ style guide's "use const whenever possible" rule.

Yeah I think "let" was a poor choice of keyword given that it's used as "const" in other languages, but we're stuck with it since it is what we want semantics-wise.

girum11 commented on src/client/app/graph/card/card.component.ts:
```diff
@@ -109,21 +106,23 @@ export class CardComponent implements OnInit {
+      // Move all associated arrows too
+      this.card.incomingArrows.asArray().forEach((inArrow: Arrow) => {
+        if (inArrow) {
+          inArrow.tipPosition = {x: newCardPosition.x, y: newCardPosition.y + this.card.size.height/2};
```
Can we change all of the object literals (throughout) to be indented consistently?

girum11 replied:
(nit) vertical space above comments?

girum11 commented on src/client/app/model/card.ts:
```diff
@@ -1,15 +1,26 @@
-  incomingArrow: Arrow;
-  outgoingArrow: Arrow;
+  // TODO(eyuelt): Change these to normal arrays since CollaborativeLists
+  // inside custom collaborative objects are not actually collaborative.
```
Imo, we should do this now. Seems harder to nest CollaborativeLists within a custom class (worse API compared to straight-up arrays), and potentially misleading.

Would only cost you a quick helper function to repeal/replace https://developers.google.com/google-apps/realtime/reference/gapi.drive.realtime.CollaborativeList.html#removeValue

eyuelt replied:
"repeal/replace" 😂

I think you might've sent me the wrong link. But I don't want to just use a helper to unwrap it here and still have all of the uses in the rest of the codebase pretend it's still a collaborative list. I want to change it everywhere. Since this would require multiple changes across the codebase, which would make merging my upcoming PR annoying, I'd rather do this in a separate PR.

girum11 commented on src/client/app/model/card.ts:
```diff
@@ -1,15 +1,26 @@
 
 export class Card extends CollaborativeObjectModel {
   protected static modelName = 'Card';
+  readonly size: Size = {width:120, height:160};
```
Indent object literals?

(I'm not a fan of C++ style declaring that vertical whitespace is the devil...)

eyuelt replied:
Meh. Alright

girum11 commented on src/client/app/graph/arrow/arrow.component.html:
```diff
@@ -1,4 +1,7 @@
-<svg:g x="0" y="0">
+<svg:g x="0"
+       y="0"
```
Let's go ahead and add the CSS rule "cursor: pointer" to the entire \<g\>

(to make it a bit easier to drag arrows)

eyuelt replied:
Done

girum11 commented on src/client/app/model/arrow.ts:
```diff
@@ -1,23 +1,24 @@
 import { CollaborativeObjectModel } from './collaborative-object-model';
 
 export class Arrow extends CollaborativeObjectModel {
   protected static modelName = 'Arrow';
```
Unused?

eyuelt replied:
The modelName field? It's used in the superclass in the registerModel function. I'm just overriding it here.

girum11 commented on src/client/app/model/arrow.ts:
```diff
@@ -1,23 +1,24 @@
-  fromCardId: string;
-  toCardId: string;
+  fromCard: Card;
+  toCard: Card;
```
Nice, I didn't know that custom objects could nest references to each other and still reload just fine.

I also like the type-safety here.

eyuelt replied:
Yeah, I'm not sure how it serializes when it goes to Realtime, but it works. Lol

girum11 commented on src/client/app/graph/canvas/canvas.service.ts:
```diff
@@ -123,6 +128,33 @@ export class CanvasService {
+      }
+      arrow.toCard = card;
+      card.incomingArrows.push(arrow);
+    } else if (arrowConnection === ArrowConnectionType.OUTGOING) {
```
Seems like this branch is never called?

Also, what's the meaning of "INCOMING" vs. "OUTGOING"? Is that enum necessary with the existing split of the arrows to "incoming" and "outgoing"?

eyuelt replied:
It's not called right now, but it will be if/when we add the ability to move the base of an arrow.
The enum just specifies how to connect the card to the arrow. I guess I could change this param to be a bool instead of using this enum. What do you think?

girum11 commented on src/client/app/graph/canvas/canvas.service.ts:
```diff
@@ -14,6 +14,11 @@ export type ViewportCoord = Point;
 // A point in the coordinate system of the canvas.
 export type CanvasCoord = Point;
 
+// Specifies the type of connection an arrow has to a card.
```
If we do need this enum, we should clarify that this enum is _relative_. See if you can clarify that these enum values could actually both be "applied" to the same arrow simultaneously (an arrow can point from one card and into another).

eyuelt replied:
Done

girum11 commented on tools/manual_typings/project/document.d.ts:
```diff
@@ -0,0 +1,3 @@
+interface Document {
+  elementsFromPoint(x: number, y: number): Element[];
```
Nice, I like the new DOM API usage here.

eyuelt replied:
![Obama - brush your shoulders off](http://stream1.gifsoup.com/view2/2287996/obama-swag-o.gif)

girum11 commented on src/client/app/graph/arrow/arrow.component.ts:
```diff
@@ -78,4 +89,70 @@ export class ArrowComponent implements OnInit {
 
+  onMousedown(event: MouseEvent) {
+    if (EventUtils.eventIsFromPrimaryButton(event)) {
+      this.tipDragging = true;
```
Just the tip?  😄 

Seems like the entire arrow is actually draggable? While we're on the subject, consider adding a TODO to invisibly increase the hitbox for an arrow?

eyuelt replied:
Smh. It's either "tip" or "head", both of which could mean other things. Lol
You're right, the mousedown listener should just be on the tip. Added a TODO.

girum11 commented on src/client/app/graph/graph.component.ts:
```diff
@@ -45,42 +45,65 @@ export class GraphComponent implements OnInit, OnDestroy {
-      card1.position = {x: 10 + numCards * 5, y: 10 + numCards * 5};
-      card1.text = 'card #' + numCards;
+      card1.position = {x: 10 + numCards * 20, y: 10 + numCards * 20};
+      card1.text = 'card #' + numCards++;
```
Ehhhh, postincrement operators in the middle of a line is surprising code, imo. I'd avoid it if possible.

eyuelt replied:
Agreed but this code is a temporary hack anyway and is getting removed in the next PR.

girum11 commented on src/client/app/graph/card/card.component.ts:
```diff
@@ -109,21 +106,23 @@ export class CardComponent implements OnInit {
+        if (outArrow) {
+          outArrow.tailPosition = { x: newCardPosition.x + this.card.size.width, y: newCardPosition.y + this.card.size.height/2 };
+          if (outArrow.toCard === null) {
+            outArrow.tipPosition = { x: newCardPosition.x + this.card.size.width + 50, y: newCardPosition.y + this.card.size.height/2 };
```
Let's follow 80 character line limits throughout?

eyuelt replied:
Done for this file. Though this gonna be hard to maintain for the whole codebase without a formatting tool that enforces this. We should probably set one up.

girum11 commented:
Overall, nicely done!

girum11 commented:
LGTM. Looks like you requested that Nati approves as well before you can merge.

eyuelt commented:
Nati said he'll read it later, so I removed him as a reviewer. It's just waiting on you now. Just putting LGTM in a comment doesn't work anymore since we removed LGTM.co. I can't remember how to actually approve though since it's been so long. Lol

girum11 commented:
Found the button.
